From 3c607a193d8fad564492ccc1ad3137be8ffe78ee Mon Sep 17 00:00:00 2001
From: z00512904 <z00512904@huawei.com>
Date: Mon, 6 Mar 2023 08:52:50 +0000
Subject: [PATCH] [Huawei] fix alloc static hugepage will fall back to other
 node

Offering: HULK
CVE: NA
Reference:
Type: bugfix
reason: When use hugetlb_alloc_hugepage() to alloc static hugepage, the page
will be allocated from other node which is different from the node user passed.
---
 mm/hugetlb.c | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/mm/hugetlb.c b/mm/hugetlb.c
index b1369941c..cda51973e 100644
--- a/mm/hugetlb.c
+++ b/mm/hugetlb.c
@@ -6271,13 +6271,13 @@ const struct hstate *hugetlb_get_hstate(void)
 EXPORT_SYMBOL_GPL(hugetlb_get_hstate);
 
 static struct page *hugetlb_alloc_hugepage_normal(struct hstate *h,
-		gfp_t gfp_mask, int nid)
+		gfp_t gfp_mask, int nid, nodemask_t *nodemask)
 {
 	struct page *page = NULL;
 
 	spin_lock(&hugetlb_lock);
 	if (h->free_huge_pages - h->resv_huge_pages > 0)
-		page = dequeue_huge_page_nodemask(h, gfp_mask, nid, NULL, NULL);
+		page = dequeue_huge_page_nodemask(h, gfp_mask, nid, nodemask, NULL);
 	spin_unlock(&hugetlb_lock);
 
 	return page;
@@ -6308,7 +6308,7 @@ struct page *hugetlb_alloc_hugepage_nodemask(int nid, int flag, nodemask_t *node
 		gfp_mask &= ~__GFP_RECLAIM;
 
 	if (flag & HUGETLB_ALLOC_NORMAL)
-		page = hugetlb_alloc_hugepage_normal(h, gfp_mask, nid);
+		page = hugetlb_alloc_hugepage_normal(h, gfp_mask, nid, nodemask);
 	else if (flag & HUGETLB_ALLOC_BUDDY)
 		page = alloc_migrate_huge_page(h, gfp_mask, nid, nodemask);
 	else
-- 
2.17.1

